storm-core/src/jvm/org/apache/storm/container/cgroup/CgroupManager.java

```diff
@@ -0,0 +1,176 @@
+import org.apache.storm.container.ResourceIsolationInterface;
+import org.apache.storm.container.cgroup.core.CpuCore;
+import org.apache.storm.container.cgroup.core.MemoryCore;
+import org.apache.storm.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class CgroupManager implements ResourceIsolationInterface {
+
+    public static final Logger LOG = LoggerFactory.getLogger(CgroupManager.class);
+
+    private CgroupCenter center;
+
+    private Hierarchy h;
+
+    private CgroupCommon rootCgroup;
+
+    private static String rootDir;
+
+    private Map conf;
+
+    public void prepare(Map conf) throws IOException {
+        this.conf = conf;
+        this.rootDir = Config.getCgroupRootDir(this.conf);
+        if (this.rootDir == null) {
+            throw new RuntimeException("Check configuration file. The supervisor.cgroup.rootdir is missing.");
+        }
+
+        File file = new File(Config.getCgroupStormHierarchyDir(conf) + "/" + this.rootDir);
+        if (!file.exists()) {
+            LOG.error("{}/{} is not existing.", Config.getCgroupStormHierarchyDir(conf), this.rootDir);
+            throw new RuntimeException("Check if cgconfig service starts or /etc/cgconfig.conf is consistent with configuration file.");
+        }
+        this.center = CgroupCenter.getInstance();
+        if (this.center == null) {
+            throw new RuntimeException("Cgroup error, please check /proc/cgroups");
+        }
+        this.prepareSubSystem(this.conf);
+    }
+
+    /**
+     * User cfs_period & cfs_quota to control the upper limit use of cpu core e.g. If making a process to fully use two cpu cores, set cfs_period_us to
+     * 100000 and set cfs_quota_us to 200000
+     */
+    private void setCpuUsageUpperLimit(CpuCore cpuCore, int cpuCoreUpperLimit) throws IOException {
+
+        if (cpuCoreUpperLimit == -1) {
+            // No control of cpu usage
+            cpuCore.setCpuCfsQuotaUs(cpuCoreUpperLimit);
+        } else {
+            cpuCore.setCpuCfsPeriodUs(100000);
+            cpuCore.setCpuCfsQuotaUs(cpuCoreUpperLimit * 1000);
+        }
+    }
+
+    public String startNewWorker(String workerId, Map resourcesMap) throws SecurityException {
+        Number cpuNum = (Number) resourcesMap.get("cpu");
+        Number totalMem = null;
+        if (resourcesMap.get("memory") != null) {
+            totalMem = (Number) resourcesMap.get("memory");
+        }
+
+        CgroupCommon workerGroup = new CgroupCommon(workerId, h, this.rootCgroup);
+        this.center.create(workerGroup);
+
+        if (cpuNum != null) {
+            CpuCore cpuCore = (CpuCore) workerGroup.getCores().get(SubSystemType.cpu);
+            try {
+                cpuCore.setCpuShares(cpuNum.intValue());
```

Comment:
ok, I agree we should test this feature to make sure idle workers can re-acquire the resources they are guaranteed once they "wake up".